src/aiida_quantumespresso/calculations/matdyn.py

```diff
@@ -47,18 +49,50 @@ def define(cls, spec):
             message='Number of kpoints in the inputs is not commensurate with those in the output')
         # yapf: enable
 
-    def _get_following_text(self):
-        """Add the kpoints after the namelist."""
-        try:
-            kpoints = self.inputs.kpoints.get_kpoints()
-        except AttributeError:
-            kpoints = self.inputs.kpoints.get_kpoints_mesh(print_list=True)
+    @staticmethod
+    def _validate_inputs(value, _):
+        """Validate the top level namespace."""
+        if 'parameters' in value:
+            parameters = value['parameters'].get_dict()
+        else:
+            parameters = {}
+
+        if parameters.get('INPUT', {}).get('flfrc', None) is not None:
+            return '`INPUT.flfrc` is set automatically from the `force_constants` input.'
+
+    def generate_input_file(self, parameters):
+        """Generate namelist input_file content given a dict of parameters.
+
+        :param parameters: 'dict' containing the fortran namelists and parameters to be used.
+          e.g.: {'CONTROL':{'calculation':'scf'}, 'SYSTEM':{'ecutwfc':30}}
+        :return: 'str' containing the input_file content a plain text.
+        """
+        kpoints = self.inputs.kpoints
+        append_string = ''
```

Comment:
Instead of initializing here, I would just do
```
file_content = super().generate_input_file(parameters)

if parameters['INPUT'].get('dos', False):
     # something
else:
    # construct append_string
    file_content += append_string

return file_content